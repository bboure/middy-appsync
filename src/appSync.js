import { GraphQlError } from './GraphQlError'
import { constant, times } from 'lodash'

const buildResponse = (response) => {
  if (response instanceof GraphQlError) {
    return {
      data: response.data,
      errorInfo: response.info,
      errorType: response.type,
      errorMessage: response.message
    }
  } else if (response instanceof Error) {
    return { errorMessage: 'Internal Server Error' }
  } else {
    return { data: response }
  }
}

export function appSync () {
  return {
    onError: async (handler, next) => {
      if (handler.error instanceof GraphQlError) {
        const response = buildResponse(handler.error)
        if (Array.isArray(handler.event)) {
          handler.response = times(handler.event.length, constant(response))
        } else {
          handler.response = response
        }

        return false
      }

      return handler.error
    },

    after: async (handler) => {
      const { response } = handler

      if (Array.isArray(handler.event)) {
        if (!Array.isArray(response) || handler.event.length !== response.length) {
          throw new Error('BatchInvoke: The response does not match the request payload')
        }

        handler.response = response.map(buildResponse)
      } else {
        handler.response = buildResponse(response)
      }
    }
  }
}
