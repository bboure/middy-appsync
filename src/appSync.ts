import { AppSyncError } from './Errors';
import middy from '@middy/core';

type AppSyncResponse = {
  data?: unknown;
  errorInfo?: unknown;
  errorType?: string;
  errorMessage?: string;
};

const buildResponse = (response: unknown): AppSyncResponse => {
  if (response instanceof AppSyncError) {
    return {
      data: response.data,
      errorInfo: response.info,
      errorType: response.type,
      errorMessage: response.message,
    };
  } else if (response instanceof Error) {
    // re-throw returned errors
    throw response;
  } else {
    return { data: response };
  }
};

export const appSync: middy.Middleware<unknown, unknown> = () => {
  return {
    onError: async (handler) => {
      const response = buildResponse(handler.error);
      if (Array.isArray(handler.event)) {
        const resp = new Array(handler.event.length);
        handler.response = resp.fill(response);
      } else {
        handler.response = response;
      }

      // mark the error as handled
      return false;
    },

    after: async (handler) => {
      const { response } = handler;
      if (Array.isArray(handler.event)) {
        if (
          !Array.isArray(response) ||
          handler.event.length !== response.length
        ) {
          throw new Error(
            'BatchInvoke: The response does not match the request payload',
          );
        }

        handler.response = response.map(buildResponse);
      } else {
        handler.response = buildResponse(response);
      }
    },
  };
};
