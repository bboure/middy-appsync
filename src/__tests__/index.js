import appSync from '../index'
import { GraphQlError } from '../error'
const middy = require('middy')

describe('appsync middleware test suite', () => {
  it('Should wrap the response in a GraphQl template', () => {
    const handler = middy((event, _, cb) => {
      cb(null, {
        field1: 'foo',
        field2: 'bar'
      })
    })

    handler.use(appSync())

    handler({}, {}, (_, response) => {
      expect(response).toMatchSnapshot()
    })
  })

  it('Should handle a GraphQlError error', () => {
    const handler = middy((event, _, cb) => {
      cb(new GraphQlError(
        'Error message',
        'Error',
        { some: 'data' },
        { info: 'value' }
      ))
    })

    handler.use(appSync())

    handler({}, {}, (_, response) => {
      expect(response).toMatchSnapshot()
    })
  })

  it('Should handle a thrown GraphQlError', () => {
    const handler = middy((event, _, cb) => {
      throw new GraphQlError(
        'Thrown Error message',
        'Thrown Error',
        { some: 'data' },
        { info: 'value' }
      )
    })

    handler.use(appSync())

    handler({}, {}, (_, response) => {
      expect(response).toMatchSnapshot()
    })
  })

  it('Should handle a GraphQlError response', () => {
    const handler = middy((event, _, cb) => {
      cb(null, new GraphQlError(
        'Returned Error message',
        'ReturnedError',
        { some: 'data' },
        { info: 'value' }
      ))
    })

    handler.use(appSync())

    handler({}, {}, (_, response) => {
      expect(response).toMatchSnapshot()
    })
  })

  it('Should handle a thrown GraphQlError with async', () => {
    const handler = middy(async (event) => {
      throw new GraphQlError(
        'Thrown Error message',
        'Thrown Error',
        { some: 'data' },
        { info: 'value' }
      )
    })

    handler.use(appSync())

    handler({}, {}, (_, response) => {
      expect(response).toMatchSnapshot()
    })
  })

  it('Should handle a returned GraphQlError with async', () => {
    const handler = middy(async (event) => {
      return new GraphQlError(
        'Returned Error message',
        'ReturnedError',
        { some: 'data' },
        { info: 'value' }
      )
    })

    handler.use(appSync())

    handler({}, {}, (_, response) => {
      expect(response).toMatchSnapshot()
    })
  })

  it('Should handle an Error', () => {
    const handler = middy((event, _, cb) => {
      cb(new Error('Some Error'))
    })

    handler.use(appSync())

    handler({}, {}, (error, response) => {
      expect(error).toBeAnInstanceof(Error)
      expect(error.message).toEqual('Some Error')
      expect(response).toBeUndefined()
    })
  })

  it('Should handle a non GraphQl Error response anonymously', () => {
    const handler = middy((event, _, cb) => {
      cb(null, new Error('Uncaught ReferenceError: myVar is not defined at index.js:123:456'))
    })

    handler.use(appSync())

    handler({}, {}, (_, response) => {
      expect(response.errorMessage).toEqual('errorMessage')
    })
  })

  it('Should not handle a non GraphQl Error', () => {
    const error = new Error('Uncaught ReferenceError: myVar is not defined at index.js:123:456')
    const handler = middy((event, _, cb) => {
      cb(error)
    })

    handler.use(appSync())

    handler({}, {}, (error, response) => {
      expect(error.message).toBe(error)
      expect(response).toBeUndefined()
    })
  })

  it('Should succeed when response matches event', () => {
    const handler = middy((event, _, cb) => cb(null, [{}, {}]))
    const middleware = appSync()
    const spy = jest.spyOn(middleware, 'after')
    handler.use(middleware)
    handler([{}, {}], {}, async (_, message) => {
      expect(spy).toHaveBeenCalledTimes(1)
      await expect(spy.mock.results[0].value).resolves.toBeUndefined()
    })
  })

  it('Should fail if the event length does not match the response length', () => {
    const handler = middy((event, _, cb) => {
      cb(null, [
        'foo',
        'bar'
      ])
    })

    const middleware = appSync()
    const spy = jest.spyOn(middleware, 'after')

    handler.use(middleware)

    handler([{}, {}, {}], {}, async (_, message) => {
      expect(spy).toHaveBeenCalledTimes(1)
      await expect(spy.mock.results[0].value).rejects.toMatchSnapshot()
    })
  })
})
