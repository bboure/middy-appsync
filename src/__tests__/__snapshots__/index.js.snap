// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appsync middleware test suite Should fail if the event length does not match the response length 1`] = `[Error: BatchInvoke: The response does not match the request payload]`;

exports[`appsync middleware test suite Should handle a GraphQlError error 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Error message",
  "errorType": "Error",
}
`;

exports[`appsync middleware test suite Should handle a GraphQlError response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Returned Error message",
  "errorType": "ReturnedError",
}
`;

exports[`appsync middleware test suite Should handle a returned GraphQlError with async 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Returned Error message",
  "errorType": "ReturnedError",
}
`;

exports[`appsync middleware test suite Should handle a thrown GraphQlError 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Thrown Error message",
  "errorType": "Thrown Error",
}
`;

exports[`appsync middleware test suite Should handle a thrown GraphQlError with async 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "errorInfo": Object {
    "info": "value",
  },
  "errorMessage": "Thrown Error message",
  "errorType": "Thrown Error",
}
`;

exports[`appsync middleware test suite Should wrap the response in a GraphQl template 1`] = `
Object {
  "data": Object {
    "field1": "foo",
    "field2": "bar",
  },
}
`;
